tools/inc/stljobs.h

```diff
@@ -305,14 +315,17 @@ struct output_collecting_pipe {
         writeHandle = create_file(pipeNameBuffer, GENERIC_WRITE | FILE_READ_ATTRIBUTES, 0, &inheritSa, OPEN_EXISTING,
             FILE_FLAG_OVERLAPPED, HANDLE{});
 
+        readEvent         = create_event(nullptr, TRUE, FALSE, nullptr);
+        overlapped.hEvent = readEvent.get();

```

Comment:
Can you point to docs indicating that such an event is necessary?

Reply:
After talking to @BillyONeal (trying to understand more of this code that he originally wrote), I would also like to understand why this explicit event is necessary.

In #1619 you said:

> Quite often it would crash upon returning from `p.wait_all()` since a kernel mode operation corrupted the stack. This goes away when using an explicit manual-reset event for `OVERLAPPED` in `output_collecting_pipe`.

From what I saw, nothing in the `OVERLAPPED` docs (or various blog posts about this API) indicates that relying on `hEvent == nullptr` could be harmful. We don't have any code that's trying to wait on the file handle.

Could it be possible that an explicit event is needed for your modifications to the parallelize tool, but not for the current sources?

Reply:
> We don't have any code that's trying to wait on the file handle.

This is most likely a problem in itself, see further below.

> Could it be possible that an explicit event is needed for your modifications to the parallelize tool, but not for the current sources?

If it does happen with the current sources, it is very rare. But it still feels lucky. The only modifications I did was using `subprocess_directive` to get a list of files beforehand. I've pushed a reduced sample to https://github.com/gix/STL/tree/gh1619-repro

Running in the parallelize directory:
```
cl -nologo -EHsc -O2 -std:c++latest -I..\inc -DWIN32_LEAN_AND_MEAN -DNOMINMAX .\parallelize.cpp -link -opt:ref -opt:icf userenv.lib
while ($true) { .\parallelize.exe "clang-format.exe -i" .; if ($LASTEXITCODE -ne 0) { echo $LASTEXITCODE; break } }
```
soon crashes here with
```
Unhandled exception thrown: write access violation. this was 0x1F6C0000120.
```
or
```
Unhandled exception at 0x00007FFD365F85D3 (ntdll.dll) in parallelize.exe: 0xC0000005: Access violation reading location 0x00000223C0000120.
```

What is very suspicious is the involvement of C0000120 which is the NTSTATUS code for `STATUS_CANCELLED`. So it looks like the kernel writes to an already free'd `OVERLAPPED` structure.

Re-reading the docs for [WaitForThreadpoolIoCallbacks](https://docs.microsoft.com/en-us/windows/win32/api/threadpoolapiset/nf-threadpoolapiset-waitforthreadpooliocallbacks) leads me to believe that my fix might even be incomplete and a call to `GetOverlappedResult` (or an equivalent wait operation) must be inserted after `CancelIoEx`. But doing so without a manual reset event just hangs. Doing several non-blocking `GetOverlappedResult` calls (without a manual reset event) shows the error switching from `ERROR_IO_INCOMPLETE` to `ERROR_OPERATION_ABORTED`.